src/reconcile/rec_visibility.c

```diff
@@ -63,67 +63,64 @@ __rec_append_orig_value(
 
     WT_ASSERT(session, upd != NULL && unpack != NULL && unpack->type != WT_CELL_DEL);
 
-    for (;; upd = upd->next) {
-        /* Done if at least one self-contained update is globally visible. */
-        if (WT_UPDATE_DATA_VALUE(upd) && __wt_txn_upd_visible_all(session, upd))
-            return (0);
+    append = tombstone = NULL;
+    total_size = 0;
 
+    /* Review the current update list, checking conditions that mean no work is needed. */
+    for (;; upd = upd->next) {
         /*
-         * If the update is restored from the history store for the rollback to stable operation we
-         * don't need the on-disk value anymore and we're done.
+         * Done if the update was restored from the history store for the rollback to stable
+         * operation.
          */
         if (F_ISSET(upd, WT_UPDATE_RESTORED_FOR_ROLLBACK))
             return (0);
 
-        /* On page value already on chain */
-        if (unpack != NULL && unpack->start_ts == upd->start_ts && unpack->start_txn == upd->txnid)
+        /* Done if the on page value already appears on the update list. */
+        if (unpack->start_ts == upd->start_ts && unpack->start_txn == upd->txnid)
+            return (0);
+
+        /*
+         * Done if at least one self-contained update is globally visible. It's tempting to pull
+         * this test out of the loop and only test the oldest self-contained update for global
+         * visibility (as visibility tests are expensive), but it's possible at lower isolation
```

Comment:
This  case not only happens at lower isolation level but also when we intentionally commit out of timestamp order.

Reply:
Fixed in 4115ae4.